### PR TITLE
Render non-reactive daily docket UI

### PR DESCRIPTION
**Pull request by slj**
Connects #2037

### CODE REVIEW

File: app/controllers/hearings/dockets_controller.rb [hunk 1]
```diff
@@ -15,7 +15,9 @@ def index
   end
 
   def show
-    redirect_to hearings_dockets_path
+    date = date_from(params[:id])
+    return not_found unless date && Judge.new(current_user).docket?(date)
+    render :index
```
Comment (slj) [inline app/controllers/hearings/dockets_controller.rb]: This handles a 1st-time direct request or reload of `hearings/dockets/YYYY-MM-DD`.

`date_from` could be part of a module down the road, but hopefully it's location in `HearingsController` will suffice for now, since I don't anticipate other apps needing to pass a date parameter in this specific format.

File: client/app/hearings/DailyDocketContainer.jsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        <div>Loading hearings, please wait...</div>
+      </div>;
+    }
+
+    if (Object.keys(this.props.dockets).length === 0) {
+      return <div>You have no upcoming hearings.</div>;
+    }
+
+    return <Docket {...this.props} />;
+  }
```
Comment (slj) [inline client/app/hearings/DailyDocketContainer.jsx]: So this container component makes the API call and renders a Docket on success, or something else.

File: client/app/hearings/Docket.jsx [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+const aodOptions = [{ value: 'grant',
+  label: 'Grant' },
+{ value: 'filed',
+  label: 'Filed' },
+{ value: 'none',
+  label: 'None' }];
+
+const getDate = (date, timezone) => {
+  return moment.tz(date, timezone).
+    format('h:mm a z').
```
Comment (slj) [inline client/app/hearings/Docket.jsx]: `moment.tz()` provided courtesy of the `moment-timezone` package.

File: client/app/hearings/Docket.jsx [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+const aodOptions = [{ value: 'grant',
+  label: 'Grant' },
+{ value: 'filed',
+  label: 'Filed' },
+{ value: 'none',
+  label: 'None' }];
+
+const getDate = (date, timezone) => {
+  return moment.tz(date, timezone).
```
Comment (slj) [inline client/app/hearings/Docket.jsx]: `moment.tz()` provided courtesy of the `moment-timezone` package.

File: client/app/hearings/index.jsx [hunk 5]
```diff
@@ -46,9 +47,18 @@ const Hearings = ({ hearings }) => {
       <BrowserRouter>
         <div>
           <Route exact path="/hearings/dockets"
-            component={() => (<DocketsContainer veteran_law_judge={hearings.veteran_law_judge} />)}/>
+            component={() => (
+              <DocketsContainer
+                veteran_law_judge={hearings.veteran_law_judge} />
+            )}
+          />
```
Comment (slj) [inline client/app/hearings/index.jsx]: Just moving things around

File: client/app/hearings/index.jsx [hunk 6]
```diff
@@ -46,9 +47,18 @@ const Hearings = ({ hearings }) => {
...
           <Route exact path="/hearings/dockets/:date"
-            component={() => (<div>Docket</div>)}/>
+            component={(props) => (
+              <DailyDocketContainer
+                veteran_law_judge={hearings.veteran_law_judge}
+                date={props.match.params.date} />
+            )}
+          />
```
Comment (slj) [inline client/app/hearings/index.jsx]: Original line was too long (broke linter 120 char rule)

File: app/models/hearing.rb [hunk 7]
```diff
@@ -29,6 +29,7 @@ def request_type
     :representative_name, \
     :appellant_last_first_mi, \
     :regional_office_name, \
+    :vbms_id, \
```
Comment (joofsh) [inline app/models/hearing.rb]: Hmm, I'm not sure this is the right approach. Are these delegations being used directly in the model, or simply for the `to_hash` on 35 below. If it's the latter, we could include the appeal object as a nested object in the to_hash. 

I fear we're going to get confused in the future and not understand where datapoints really come from if we delegate everything.

File: client/app/hearings/Docket.jsx [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+  return moment().
+    format('h:mm a').
+    replace(/(p|a)m/, '$1.m.');
+};
+
+export class Docket extends React.Component {
+
+  componentDidMount = () => {
+    // TEMP logic to show Saving.../Last saved at <time>
+    setInterval(() => {
```
Comment (joofsh) [inline client/app/hearings/Docket.jsx]: This is outside the scope of this PR. Can we remove it please

File: client/test/karma/hearings/DailyDocketContainer-test.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+const store = createStore(hearingsReducers, { dockets: {} }, applyMiddleware(thunk));
+
+/* eslint-disable camelcase */
+/* eslint-disable no-unused-expressions */
+/* eslint-disable max-statements */
+describe('DailyDocketContainer', () => {
+  let wrapper;
+
+  spy(ApiUtil, 'get');
```
Comment (joofsh) [inline client/test/karma/hearings/DailyDocketContainer-test.js]: This is great. Unfortunately though there's already a helper for this. Check out `ApiUtilStub.js`

File: client/test/karma/hearings/DailyDocketContainer-test.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+import hearingsReducers from '../../../app/hearings/reducers/index';
+import { populateDockets } from '../../../app/hearings/actions/Dockets';
+import DailyDocketContainer from '../../../app/hearings/DailyDocketContainer';
+
+const store = createStore(hearingsReducers, { dockets: {} }, applyMiddleware(thunk));
+
+/* eslint-disable camelcase */
+/* eslint-disable no-unused-expressions */
+/* eslint-disable max-statements */
+describe('DailyDocketContainer', () => {
```
Comment (joofsh) [inline client/test/karma/hearings/DailyDocketContainer-test.js]: These tests look great 👍

File: app/assets/stylesheets/_hearings.scss [hunk 11]
```diff
@@ -174,9 +178,158 @@ body {
...
+    .meta {
+      font-size: inherit;
+    }
+  }
+
+  .cf-hearings-docket {
+    margin-top: 0;
+
+    // TODO: see if this can become a part of the general style
+    .Select-arrow-zone {
```
Comment (slj) [inline app/assets/stylesheets/_hearings.scss]: As I understand it, the dropdowns here are based off an NPM package that used these classnames.  I'm just overwriting the original rules.

File: app/controllers/hearings_controller.rb [hunk 12]
```diff
@@ -32,5 +32,16 @@ def verify_access
   def set_application
     RequestStore.store[:application] = "hearings"
   end
+
+  def date_from(param)
```
Comment (slj) [inline app/controllers/hearings_controller.rb]: Will rename

File: client/app/hearings/Docket.jsx [hunk 13]
```diff
@@ -0,0 +1,180 @@
+import React from 'react';
```
Comment (slj) [inline client/app/hearings/Docket.jsx]: I will rename the child component to DailyDocket.

File: client/app/hearings/Docket.jsx (hunk 8 above)
Comment (slj) [inline client/app/hearings/Docket.jsx]: The text is part of the AC.  The temporary logic gives our stakeholders/testers a convenient way to see all of the text (Saving.../Last saved at...) in action.  The plan is to remove it when the save functionality is actually implemented.

Comment (joofsh) [inline client/app/hearings/Docket.jsx]: Okay looks like this was a misunderstanding between @lakohl and I. I explicitly had a comment in the original ticket to not do the saving interface

> No need to implement the saving interface

Going forward, please check the technical notes section before implementing something like this.

File: app/controllers/hearings/dockets_controller.rb [hunk 14]
```diff
@@ -15,7 +15,9 @@ def index
   end
 
   def show
-    redirect_to hearings_dockets_path
+    date = date_from(params[:id])
+    return not_found unless date && Judge.new(current_user).docket?(date)
```
Comment (joofsh) [inline app/controllers/hearings/dockets_controller.rb]: One reason is the `Judge` object caches its `upcoming_dockets`. So if you re-instantiate `Judge.new()` twice, you will make 2 DB requests looking up hearings. Having a ` def judge` scales better as more code gets added to this controller, and as the judge model becomes more complex

File: app/models/hearing.rb (hunk 7 above)
Comment (slj) [inline app/models/hearing.rb]: When sending responses with hashed objects, it's easy to transfer unnecessary data (hearing prep is doing so).  For this ticket, the hash didn't need everything from an appeal, just the items in lines 29-32.

I don't see how this becomes confusing when it's clear these methods come from `:appeal`

To be fair, there may be a better way, but I think the necessary change(s) will become clearer as more of the app is fleshed out and things get optimized, so I hope you can tolerate this for now.

File: client/test/karma/hearings/DailyDocketContainer-test.js (hunk 9 above)
Comment (slj) [inline client/test/karma/hearings/DailyDocketContainer-test.js]: I could not get `ApiUtilStub` to work.  I tried for a very long time before going with the approach shown.

File: client/app/hearings/Docket.jsx [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+        document.getElementsByClassName('saving')[0].innerHTML = `Last saved at ${now()}`;
+      } else {
+        document.getElementsByClassName('saving')[0].innerHTML = 'Saving...';
+      }
+    }, 3000);
+  }
+
+  render() {
+
+    let hearings = this.props.dockets[this.props.date].hearings_hash;
```
Comment (slj) [inline client/app/hearings/Docket.jsx]: OK

File: app/models/judge.rb [hunk 16]
```diff
@@ -19,4 +23,10 @@ def upcoming_hearings_grouped_by_date
   def upcoming_hearings
     Hearing.repository.upcoming_hearings_for_judge(user.vacols_id).sort_by(&:date)
   end
+
+  def upcoming_hearings_on(date)
+    upcoming_hearings.select do |hearing|
+      hearing.date.between?(date, date.end_of_day)
```
Comment (slj) [inline app/models/judge.rb]: Agreed.

File: app/models/hearing.rb (hunk 7 above)
Comment (joofsh) [inline app/models/hearing.rb]: Hmm, I'm not sure what you mean. I agree we shouldn't send unnecessary data to the frontend. If we are doing that elsewhere we should fix it.

It's not confusing to you because you know this code well :) A new developer coming in and looking at the frontend code will see a flattened object for the hearings and will think this `vbms_id` is part of the hearing record when it is in fact the vbms_id for the veteran/appeal. We're basically referencing this in a different way here than we did in Dispatch. `task.appeal.vbms_id`. Inconsistencies = confusion.

File: client/test/karma/hearings/DailyDocketContainer-test.js (hunk 9 above)
Comment (joofsh) [inline client/test/karma/hearings/DailyDocketContainer-test.js]: can you be more specific on what didn't work? That stub util is used heavily throughout Reader tests. We should be able to copy the same pattern. I don't want our code to diverge on testing patterns here

File: spec/feature/hearings_spec.rb [hunk 17]
```diff
@@ -72,6 +72,11 @@
       visit "/hearings/dockets"
       expect(page).to have_content("Page not found")
     end
+
+    scenario "Shows a daily docket" do
+      visit "/hearings/dockets/2017-01-05"
+      expect(page).to have_content("Daily Docket")
```
Comment (joofsh) [inline spec/feature/hearings_spec.rb]: Can we test a little more than just the daily docket header loading? Do you mind adding a test that at least 1 "hearing row" showed up

File: app/models/judge.rb [hunk 18]
```diff
@@ -10,6 +10,10 @@ def upcoming_dockets
     end
   end
 
+  def docket?(date)
+    return true if upcoming_hearings_on(date).count > 0
```
Comment (joofsh) [inline app/models/judge.rb]: minor nit, you can simply this to just be the conditional.  `upcoming_hearings_on(date).count > 0` that inherently returns true

File: app/controllers/hearings/dockets_controller.rb [hunk 19]
```diff
@@ -15,7 +15,9 @@ def index
   end
 
   def show
-    redirect_to hearings_dockets_path
+    date = date_from_string(params[:id])
+    return not_found unless date && Judge.new(current_user).docket?(date)
```
Comment (joofsh) [inline app/controllers/hearings/dockets_controller.rb]: One reason is the Judge object caches its upcoming_dockets. So if you re-instantiate Judge.new() twice, you will make 2 DB requests looking up hearings. Having a def judge scales better as more code gets added to this controller, and as the judge model becomes more complex

File: app/assets/stylesheets/_hearings.scss [hunk 20]
```diff
@@ -174,9 +178,158 @@ body {
...
+      background-position: center center;
```
Comment (mmlumba) [inline app/assets/stylesheets/_hearings.scss]: typo?

File: client/test/karma/hearings/DailyDocketContainer-test.js (hunk 9 above)
Comment (slj) [inline client/test/karma/hearings/DailyDocketContainer-test.js]: The test kept failing because the expectation `....get.calledWith(....).to.be.true` was consistently coming back false.  I tried for a very long time to use the `ApiUtilStub` pattern and could not succeed.  I'm not against changing this but it's low on my list of things to get done.

File: app/models/judge.rb (hunk 18 above)
Comment (slj) [inline app/models/judge.rb]: OMG of course.  Thank you.

File: app/controllers/hearings/dockets_controller.rb (hunk 19 above)
Comment (slj) [inline app/controllers/hearings/dockets_controller.rb]: Done.

File: app/assets/stylesheets/_hearings.scss (hunk 20 above)
Comment (slj) [inline app/assets/stylesheets/_hearings.scss]: Changing to `background-position: center`

File: client/test/karma/hearings/DailyDocketContainer-test.js (hunk 9 above)
Comment (joofsh) [inline client/test/karma/hearings/DailyDocketContainer-test.js]: I think it's important that we keep our code consistent if at all possible. Can you try one more time please following the example from Reader here: https://github.com/department-of-veterans-affairs/caseflow/blob/master/client/test/node/reader/DecisionReviewer-test.js#L199-L202

I think if you reference `ApiUtilStub.apiGet` rather than `ApiUtil.get` it should work

File: spec/feature/hearings_spec.rb (hunk 17 above)
Comment (slj) [inline spec/feature/hearings_spec.rb]: Done.

File: client/test/karma/hearings/DailyDocketContainer-test.js [hunk 21]
```diff
@@ -0,0 +1,81 @@
+import React from 'react';
```
Comment (joofsh) [inline client/test/karma/hearings/DailyDocketContainer-test.js]: Running these tests locally I see a lot of spam in the logs. I think it should be a simple typo fix?

screen shot 2017-07-10 at 6 24 53 pm

Comment (slj) [inline client/test/karma/hearings/DailyDocketContainer-test.js]: Could have sworn I've seen this before, even before I started adding tests.  However, I will go back to ensure my code is not the culprit.  Thx

Comment (slj) [inline client/test/karma/hearings/DailyDocketContainer-test.js]: I got no results from this search in `client/app/hearings`:

`find . -name '*.js*' -print | xargs \grep 'oneOfType'`

I'm confident that the issue lies outside the code in this PR.  I definitely make an issue for this.

File: client/test/karma/hearings/DailyDocketContainer-test.js (hunk 9 above)
Comment (slj) [inline client/test/karma/hearings/DailyDocketContainer-test.js]: Done.

File: client/test/karma/hearings/DailyDocketContainer-test.js (hunk 21 above)
Comment (joofsh) [inline client/test/karma/hearings/DailyDocketContainer-test.js]: okay sounds good, thanks! 👍
